### PR TITLE
EOS-16026: Disruptive Upgrade: Perform post upgrade

### PR DESCRIPTION
**Pull request by Madhura-08**
- Check if there are any resources in the cluster
  If yes, exit.
- Check if cluster is in standby mode. If not, make standby mode ON
- Load new config (Merge old and new if required)
- Create pacemaker resources
- Unstandby the cluster

Signed-off-by: Madhura Mande <madhura.mande@seagate.com>

## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    EOS-16026: Disruptive Upgrade: Perform post upgrade  
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Provide functionality to perform post upgrade
  </code>
</pre>
## Solution
<pre>
  <code>
    A set of routines packed in a python script which will do the post job. Assuming some pre-requisites in place.
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    - perform post-upgrade with all the resources in place. Script should exit.
    - perform post-upgrade when cluster is not in standby mode.
    - perform post-upgrade when there are no resources and cluster is in standby mode.
  </code>
</pre>

### CODE REVIEW

**Comment by indrajitzagade** [comment]
@Madhura-08 Please update the PR template.

File: ha/setup/post_disruptive_upgrade.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+from cortx.utils.log import Log
+from ha.execute import SimpleCommand
+from ha.const import (BACKUP_DEST_DIR_CONF, CONFIG_DIR)
+from ha.setup.create_pacemaker_resources import create_all_resources
+
+
+def check_for_any_resource_presence() -> None:
+    '''Check if any resources are already present in a cluster.
+       if yes, means, pre-upgrade steps failes hence exit.'''
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:26]: *failed

File: ha/setup/post_disruptive_upgrade.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    Log.info('Check for any resource presence in a cluster')
+
+    list_pcs_resources = 'crm_resource --list-raw'
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:30]: Can this be moved to const.py?

File: ha/setup/post_disruptive_upgrade.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    resource_list = SimpleCommand().run_cmd(list_pcs_resources)
+
+    resource_list.split('\n')
+
+    if resource_list:
+         Log.error('Some resources are already present. \
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:36]: will performing the Post-Upgrade again will help resolve this issue?
If not, I think this error log should provide the correct next step.

File: ha/setup/post_disruptive_upgrade.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                    Perform Post-Upgrade again')
+        return
+
+def is_cluster_standby_on() -> None:
+    '''Check if cluster is in standby mode. If not, make standby mode ON'''
+
+    Log.info('Check cluster is in standby mode')
+    check_standby_mode = 'crm_standby --query | awk \'{print $3}\' | cut -d \'=\' -f 2'
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:44]: I think the command portion could go to const.py file.

File: ha/setup/post_disruptive_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    value = SimpleCommand().run_cmd(check_standby_mode)
+
+    if value.lower() != 'on':
+        Log.error('cluster is not in standby mode. Making standby mode ON')
+        standby_cmd = "pcs node standby --all --wait=600"
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:49]: Already we have a "PCS_NODE_STANDBY" command defined in const.py file, please use that.

Comment (indrajitzagade) [reply ha/setup/post_disruptive_upgrade.py:49]: Already we have a "BASE_WAIT_TIME" defined in const.py file for the sleep purpose, please check how it is used in other standby cases and apply it here.

File: ha/setup/post_disruptive_upgrade.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        SimpleCommand().run_cmd(check_standby_mode)
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:50]: *standby_cmd

File: ha/setup/post_disruptive_upgrade.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+def restore_config(src_dir: str = BACKUP_DEST_DIR_CONF, dest_dir: str = CONFIG_DIR) -> None:
+    '''Restores the config taken at the pre-upgrade step'''
+
+    # TODO: Merging of old and new config if required
+    try:
+        if os.path.exists(src_dir) and os.listdir(src_dir):
+            copytree(src_dir, dest_dir)
+    except Exception as e:
+        pass
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:66]: why we need pass here?

File: ha/setup/post_disruptive_upgrade.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    Log.info('Restoring the consul backup')
+    # TODO
+
+def restore_config(src_dir: str = BACKUP_DEST_DIR_CONF, dest_dir: str = CONFIG_DIR) -> None:
+    '''Restores the config taken at the pre-upgrade step'''
+
+    # TODO: Merging of old and new config if required
+    try:
+        if os.path.exists(src_dir) and os.listdir(src_dir):
+            copytree(src_dir, dest_dir)
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:64]: where does this copytree() come from?

Comment (indrajitzagade) [reply ha/setup/post_disruptive_upgrade.py:64]: Also, what if the dest_dir (and contents) already exists? Will this still work?

File: ha/setup/post_disruptive_upgrade.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+def create_resources() -> None:
+    '''create required resources'''
+    Log.info('Creating the resources in the cluster')
+    create_all_resources()
+
+def _unstandby_cluster() -> None:
+    '''Unstandby the cluster'''
+
+    unstandby_cmd = "pcs node unstandby --all --wait=600"
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:77]: Similar comments as "standby" above, these are present in const.py file.

File: ha/setup/post_disruptive_upgrade.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        if os.path.exists(src_dir) and os.listdir(src_dir):
+            copytree(src_dir, dest_dir)
+    except Exception as e:
+        pass
+        Log.error(f'post upgrade failed at config restore phase: {e}')
+
+def create_resources() -> None:
+    '''create required resources'''
+    Log.info('Creating the resources in the cluster')
+    create_all_resources()
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:72]: This call needs some arguments, please check the call in ha_setup.py file.

**Comment by indrajitzagade** [comment]
Overall please comment on how this post-upgrade will be called.

**Comment by azheregelya** [comment]
Could you please improve PR template describing the problem and solution?

File: ha/setup/post_disruptive_upgrade.py (hunk 3 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:36]: I think, there is no going back in the upgrade process. If something fails, we have to start the process from the scratch. So, if resources are present, means, pre-upgrade did not succeed. Means, we need to perform pre-upgrade again. Will update this in the message.

File: ha/setup/post_disruptive_upgrade.py (hunk 1 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 2 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Yes. Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 4 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Yes. Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 5 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Ok. Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 6 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:40]: Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 7 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Now, I am raising the exception with proper message

File: ha/setup/post_disruptive_upgrade.py (hunk 8 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: I am using copystat now from the shutil package. copytree will not work if dest dir already exists but copystat is.

File: ha/setup/post_disruptive_upgrade.py (hunk 9 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 10 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:60]: Yeah. Check that. Thanks

File: ha/setup/ha_setup.py [hunk 11]
```diff
@@ -125,6 +125,28 @@ def get_node_name(self):
         Log.info(f"Read node name: {node_name}")
         return node_name
 
+    @staticmethod
```
Comment (ajaysrivas) [inline ha/setup/ha_setup.py:128]: How is this change related to post_upgrade?

File: ha/setup/post_disruptive_upgrade.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        Log.error('cluster is not in standby mode. Making standby mode ON')
+        standby_cmd = PCS_CLUSTER_STANDBY + f' --wait={CLUSTER_STANDBY_UNSTANDBY_TIMEOUT}'
+        SimpleCommand().run_cmd(standby_cmd)
+        Log.info('cluster standby mode on success')
+
+def restore_consul_backup():
+
+    Log.info('Restoring the consul backup')
+
+def restore_config() -> None:
```
Comment (ajaysrivas) [inline ha/setup/post_disruptive_upgrade.py:48]: The old and new config needs to be merged, restore is just one scenario.

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

File: ha/const.py [hunk 13]
```diff
@@ -131,6 +134,7 @@
 CLUSTER_RETRY_COUNT = 6
 BASE_WAIT_TIME = 5
 NODE_STOP_TIMEOUT = 300 # 300 sec to stop single node
+CLUSTER_STANDBY_UNSTANDBY_TIMEOUT = 600 # 300 sec to stop single node
```
Comment (indrajitzagade) [inline ha/const.py:137]: Please update the comment.

File: ha/setup/ha_setup.py [hunk 14]
```diff
@@ -125,6 +125,28 @@ def get_node_name(self):
...
+        """
+        try:
+            print(Conf)
+            s3_instances = Conf.get(Cmd._index, f"server_node.{machine_id}.s3_instances")
+            if int(s3_instances) < 1:
+                raise HaConfigException(f"Found {s3_instances} which is invalid s3 instance count.")
+            return int(s3_instances)
+        except Exception as e:
+            Log.error(e)
+            # Log.error(f"Found {s3_instances} which is invalid s3 instance count. Error: {e}")
```
Comment (indrajitzagade) [inline ha/setup/ha_setup.py:147]: Please delete if not needed.

File: ha/setup/ha_setup.py [hunk 15]
```diff
@@ -381,17 +403,20 @@ class UpgradeCmd(Cmd):
     """
     name = "upgrade"
 
-    def __init__(self, args):
+    def __init__(self, args=None):
         """
         Init method.
         """
         super().__init__(args)
+        # self._config = ConfigCmd(args)
```
Comment (indrajitzagade) [inline ha/setup/ha_setup.py:411]: Please delete if not needed.

File: ha/setup/ha_setup.py (hunk 11 above)
Comment (indrajitzagade) [reply ha/setup/ha_setup.py:128]: There is an existing _get_s3_instance() method at line #314 below.

**Review by indrajitzagade — CHANGES_REQUESTED** [review]
Please fix the comments.

File: ha/setup/ha_setup.py [hunk 16]
```diff
@@ -381,17 +403,20 @@ class UpgradeCmd(Cmd):
...
+        machine_id = self.get_machine_id()
+        self.s3_instance = Cmd.get_s3_instance(machine_id)
```
Comment (indrajitzagade) [inline ha/setup/ha_setup.py:413]: We cannot use confstore during run-time.
cc @ajaysrivas

File: ha/setup/ha_setup.py (hunk 11 above)
Comment (Madhura-08) [reply ha/setup/ha_setup.py:128]: Its a private method of ConfigCmd class. And private methods, we can not access outside (though, python will allow this in our case). 2nd, now, this method is not just required at config phase, but it has use case at upgrade case too. 
And, the instance count will be same (static). Right ?
Hence, I made it static method of Cmd class. implementation wise, its same as _get_s3_instance().
Let me know if still change is needed. Will do it.

File: ha/setup/ha_setup.py (hunk 14 above)
Comment (Madhura-08) [reply ha/setup/ha_setup.py:147]: Its needed. Hence uncommented.

File: ha/setup/ha_setup.py (hunk 15 above)
Comment (Madhura-08) [reply ha/setup/ha_setup.py:411]: Done

File: ha/setup/ha_setup.py (hunk 16 above)
Comment (Madhura-08) [reply ha/setup/ha_setup.py:413]: I think we already discussed this.

File: ha/setup/ha_setup.py [hunk 17]
```diff
@@ -323,24 +343,24 @@ def process(self):
...
-        """
-        try:
-            s3_instances = Conf.get(self._index, f"server_node.{machine_id}.s3_instances")
-            if int(s3_instances) < 1:
-                raise HaConfigException(f"Found {s3_instances} which is invalid s3 instance count.")
-            return int(s3_instances)
-        except Exception as e:
-            Log.error(f"Found {s3_instances} which is invalid s3 instance count. Error: {e}")
-            raise HaConfigException(f"Found {s3_instances} which is invalid s3 instance count.")
+    #def _get_s3_instance(self, machine_id: str) -> int:
```
Comment (Madhura-08) [inline ha/setup/ha_setup.py:346]: For now I kept this commented. Will remove or uncomment as per the need.

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

**Comment by gowthamchinna** [comment]
test this please

File: ha/setup/ha_setup.py (hunk 17 above)
Comment (indrajitzagade) [reply ha/setup/ha_setup.py:346]: Please remove before the merge if it is not needed.

File: ha/setup/post_disruptive_upgrade.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+from deepdiff import DeepDiff
+from cortx.utils.log import Log
+from ha.core.error import UpgradeError
+from ha.execute import SimpleCommand
+from ha.const import (BACKUP_DEST_DIR_CONF, CONFIG_DIR, SOURCE_CONFIG_PATH,
+                        RA_LOG_DIR, CHECK_PCS_STANDBY_MODE, \
+                        CLUSTER_STANDBY_UNSTANDBY_TIMEOUT, PCS_CLUSTER_STANDBY, \
+                        PCS_CLUSTER_UNSTANDBY)
+from ha.setup.create_pacemaker_resources import create_all_resources
+import yaml
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:32]: Please move this above at line #19 (after import os)

File: ha/setup/post_disruptive_upgrade.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        _switch_cluster_mode(PCS_CLUSTER_STANDBY)
+    Log.info('#### All post-upgrade prerequisites are in place ####')
+
+def _restore_consul_backup():
+
+    Log.info('Restoring the consul backup')
+
+def _yaml_to_dict(yaml_file=None):
+    '''
+       Convert yaml format key value info i the form of
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:75]: 'i'?

File: jenkins/pyinstaller/v2/requirements.txt [hunk 20]
```diff
@@ -6,3 +6,4 @@ dataclasses==0.7
 python-consul==1.1.0
 pathlib==1.0.1
 pytest==5.4.2
+deepdiff==5.5.0
```
Comment (indrajitzagade) [inline jenkins/pyinstaller/v2/requirements.txt:9]: I think we cannot simply add a third party tool, there is an approval process.
But lets check if we absolutely need this third party tool, we should try to achieve with currently available tools.

File: ha/setup/post_disruptive_upgrade.py (hunk 18 above)
Comment (ajaysrivas) [reply ha/setup/post_disruptive_upgrade.py:32]: @Madhura-08 Lets not use any new python package. Please check if a package can be used from https://github.com/Seagate/cortx-re/blob/main/scripts/third-party-rpm/python-requirements.txt. If not, then we should write our own code.

File: jenkins/pyinstaller/v2/requirements.txt (hunk 20 above)
Comment (ajaysrivas) [reply jenkins/pyinstaller/v2/requirements.txt:9]: @Madhura-08 Lets not use any new python package. Please check if a package can be used from https://github.com/Seagate/cortx-re/blob/main/scripts/third-party-rpm/python-requirements.txt. If not, then we should write our own code.

Comment (Madhura-08) [reply jenkins/pyinstaller/v2/requirements.txt]: Sure. Modified the logic

File: ha/setup/post_disruptive_upgrade.py (hunk 19 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:74]: Sorry typing mistake. It should be in

File: ha/setup/post_disruptive_upgrade.py (hunk 18 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: I think its not a standard package. Its a third party package. Hence not imported with standard imports according to pylint suggestion

File: ha/setup/ha_setup.py (hunk 17 above)
Comment (Madhura-08) [reply ha/setup/ha_setup.py:346]: Sure.

File: ha/setup/post_disruptive_upgrade.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+    '''
+       Load the new config at proper location after the
+       RPM upgrade as part of post-upgrade process
+    '''
+
+    dest_dir = CONFIG_DIR
+    new_src_dir = SOURCE_CONFIG_PATH
+
+    HA_BACKUP_CONF_FILE = BACKUP_DEST_DIR_CONF + '/' + 'ha.conf'
+    HA_SOURCE_CONF = SOURCE_CONFIG_PATH + '/' + 'ha.conf'
```
Comment (ajaysrivas) [inline ha/setup/post_disruptive_upgrade.py:94]: Please do not hardcode these files. Pass these as argument to perform_post_upgrade function and keep the default value to the arguments as above.

File: ha/setup/post_disruptive_upgrade.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+
+    # Convert yaml to dictionary
+    old_backup_conf_dict = _yaml_to_dict(HA_BACKUP_CONF_FILE)
+    new_conf_dict = _yaml_to_dict(HA_SOURCE_CONF)
+
+    # convert dictionary to set data structure
+    old_conf_set = set(old_backup_conf_dict)
+    new_conf_set = set(new_conf_dict)
```
Comment (ajaysrivas) [inline ha/setup/post_disruptive_upgrade.py:102]: Does it work for multi level dictionary?

File: ha/setup/post_disruptive_upgrade.py (hunk 21 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:94]: Ok. Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 22 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:102]: I think No. 
Just tried an example.
>>> a = {'ppp': {'qq': 1, 'rr': 2}}
>>> set(a)
{'ppp'}
>>>

Comment (ajaysrivas) [reply ha/setup/post_disruptive_upgrade.py:102]: Then implementation needs to be changed to handle this scenario.

File: ha/setup/ha_setup.py (hunk 17 above)
Comment (indrajitzagade) [reply ha/setup/ha_setup.py:351]: Waiting for final patch where this is removed.

File: ha/setup/post_disruptive_upgrade.py (hunk 12 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Handled in a different way now. Please check.

File: ha/setup/ha_setup.py (hunk 17 above)
Comment (Madhura-08) [reply ha/setup/ha_setup.py]: Done. Removed.

File: ha/setup/post_disruptive_upgrade.py (hunk 22 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Used different way now. Please check.

File: ha/setup/post_disruptive_upgrade.py (hunk 19 above)
Comment (indrajitzagade) [reply ha/setup/post_disruptive_upgrade.py:74]: Please fix.

File: ha/setup/post_disruptive_upgrade.py [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+
+    # Convert yaml to dictionary
+    old_backup_conf_dict = _yaml_to_dict(ha_backup_conf)
+    new_conf_dict = _yaml_to_dict(ha_source_conf)
+
+    # Note: There are 3 scenarios for conf file upgrade
+    # 1. New conf key-value pair can be introduced
+    # 2. Already present conf key can be updated with new value
+    # 3. A conf key-value can be deleted
+    # Here, we are considering or assuming thatthere will not be updation.
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:113]: *that there

File: ha/setup/post_disruptive_upgrade.py [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+    # Upgrade means a new key value will be added. 2nd scenario can be handled
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:114]: Is the 2nd scenario handled here? I think the comment should dictate that clearly.

**Review by indrajitzagade — COMMENTED** [review]
Some minor comments are open.

File: ha/setup/post_disruptive_upgrade.py (hunk 23 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Done.

File: ha/setup/post_disruptive_upgrade.py (hunk 24 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:114]: No.. Please check now.

File: ha/setup/post_disruptive_upgrade.py (hunk 19 above)
Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py]: Done.

File: ha/setup/post_disruptive_upgrade.py [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+    _switch_cluster_mode(PCS_CLUSTER_UNSTANDBY)
+    Log.info('### cluster is up and running ###')
+
+def perform_post_upgrade(s3_instances=None):
+    '''Starting routine for post-upgrade process'''
+    Log.init(service_name="post_disruptive_upgrade", log_path=RA_LOG_DIR, level="INFO")
+    _check_for_any_resource_presence()
+    _is_cluster_standby_on()
+    _load_config()
+    _create_resources(s3_instances)
```
Comment (indrajitzagade) [inline ha/setup/post_disruptive_upgrade.py:162]: Will this be called on all nodes?

Comment (Madhura-08) [reply ha/setup/post_disruptive_upgrade.py:162]: When I created the LLD for this feature, it was suppose to be for single node VM. Created LLD assuming the same fact. The previous ticket was also tested in same way.
So, I am assuming it is for single node.
Please let me know if there are any concerns

**Review by ajaysrivas — APPROVED** [review]

File: ha/setup/post_disruptive_upgrade.py (hunk 25 above)
Comment (indrajitzagade) [reply ha/setup/post_disruptive_upgrade.py:162]: ok

**Review by indrajitzagade — APPROVED** [review]
